xmtp_mls/src/storage/encrypted_store/db_connection.rs

```diff
@@ -19,37 +26,75 @@ pub type DbConnection = DbConnectionPrivate<sqlite_web::connection::WasmSqliteCo
 // callers should be able to accomplish everything with one conn/reference.
 #[doc(hidden)]
 pub struct DbConnectionPrivate<C> {
-    inner: Arc<Mutex<C>>,
+    // Connection with read-only privileges
+    read: Option<Arc<Mutex<C>>>,
+    // Connection with write privileges
+    write: Arc<Mutex<C>>,
+    // Is any connection (possibly this one) currently in a transaction?
+    transaction_lock: Arc<Mutex<()>>,
+    // Is this particular connection in a transaction?
+    in_transaction: Arc<AtomicBool>,
 }
 
 /// Owned DBConnection Methods
 impl<C> DbConnectionPrivate<C> {
     /// Create a new [`DbConnectionPrivate`] from an existing Arc<Mutex<C>>
-    pub(super) fn from_arc_mutex(conn: Arc<Mutex<C>>) -> Self {
-        Self { inner: conn }
+    pub(super) fn from_arc_mutex(
+        write: Arc<Mutex<C>>,
+        read: Option<Arc<Mutex<C>>>,
+        transaction_lock: Arc<Mutex<()>>,
+    ) -> Self {
+        Self {
+            read,
+            write,
+            transaction_lock,
+            in_transaction: Arc::new(AtomicBool::new(false)),
+        }
     }
 }
 
 impl<C> DbConnectionPrivate<C>
 where
     C: diesel::Connection,
 {
+    pub(crate) fn start_transaction<Db: XmtpDb<Connection = C>>(
+        &self,
+    ) -> Result<TransactionGuard<'_>, StorageError> {
+        let guard = self.transaction_lock.lock();
```

Comment:
what can go wrong if we don't have this mutex?

Reply:
Transactions will get nested.

Reply:
do we have a test that fails with nested transactions -- shouldn't we avoid nested transactions altogether now that we no longer have async? 

Or, if we have nested transaction in sync code, is this Mutex just covering up for a suboptimal code design somewhere else in libxmtp that we need to investigate

Reply:
Transactions or other writes could get nested if another provider tries to start a transaction before this transaction closes.

Reply:
> shouldn't we avoid nested transactions altogether now that we no longer have async?

With the lib itself being async a worker could try to start a write while this thread is still writing. This mutex will block that.

Reply:
> do we have a test that fails with nested transactions

We have a test that checks for nested transactions here: https://github.com/xmtp/libxmtp/blob/17a8d804ff8fee4464cb7db71911cdbf7e4142c5/xmtp_mls/src/groups/mod.rs#L3801